### PR TITLE
[18.0][ADD] mrp_workorder_subcontracting

### PR DESCRIPTION
**Pull request by aleboldro02**
This PR introduces `mrp_workorder_subcontracting`, a new module to manage
subcontracting flows at manufacturing workorder level.

## Goal

Standard Odoo subcontracting is mainly managed at bom and manufacturing order level.
This module supports scenarios where only specific routing operations need to
be subcontracted, while keeping traceability between manufacturing orders, workorders, purchase documents and stock transfers.

## Main features

- Add subcontracting configuration on routing operations.
- Copy subcontracting configuration from routing operations to generated work orders.
- Allow subcontracting configuration directly on work orders before documents are created.
- Assign one or more work orders to subcontractors through a dedicated wizard.
- Support allowed subcontractors and subcontracting service products per operation/work order.
- Link purchase order lines, delivery moves and return moves to the related work order.
- Add smart buttons and list views to navigate subcontracting documents from MOs, POs and pickings.
- Track subcontracting status on work orders.

### Standard purchase flow

The wizard can create new subcontracting RFQs or reuse an existing draft RFQ.

When the purchase order is confirmed, the module creates the subcontracting
outgoing transfer and then the incoming transfer. Purchase order types define
which subcontracting operation types are used.

The module also supports an immediate-return variant, where the incoming
transfer is prepared at purchase confirmation.

### Multi-supplier bidding flow

When multiple suppliers are selected, the wizard creates one RFQ per supplier.

Confirming one RFQ opens a winning bid confirmation wizard. After confirmation,
competing RFQs are either cancelled or their losing lines are set to zero
quantity and locked.

### Urgent flow

The urgent flow creates outgoing subcontracting transfers directly, without
requiring a purchase order first.

An urgency reason is required and is posted on the related manufacturing order.

### Subcontractor stock flow

The subcontractor-stock flow creates the incoming transfer for processed goods
available from the subcontractor side.

A purchase order can optionally be linked when a single supplier is selected.

## Logistics and completion

The module supports two supply methods:

- sending raw materials/components to the subcontractor;
- sending/receiving the finished product through virtual subcontracting locations.

When subcontracting receipts are completed, the related work order is evaluated
and completed for the received quantity. If logistics documents are closed
without a valid receipt, the work order is marked with a subcontracting
exception.

## Configuration

The module creates default subcontracting locations, stock operation types and
purchase order types.

After installation, warehouses and suppliers must be configured with the
appropriate subcontracting picking types and supplier-specific locations to
keep stock traceability per subcontractor.

Detailed setup instructions are included in `readme/CONFIGURE.md`.

## Status

Alpha.

The module is functional but still in early development. The main remaining
work is to add automated tests and further validate edge cases around partial
receipts, backorders, multi-company and multi-warehouse setups.

### CODE REVIEW

**Comment by sergiocorato** [comment]
@aleboldro02 thanks for the proposal! This seems a huge implementation of a simpler way to do subcontracting, I'll check it.

@OCA/manufacturing-maintainers what is needed to approve the workflow run?

**Comment by ivantodorovich** [comment]
approved

**Review by andreampiovesana — APPROVED** [review]
LGTM

**Comment by sergiocorato** [comment]
> approved

Hi @ivantodorovich , to approve in a way to be counted, it must be used the button in tab "Files changed" to submit a review.

**Comment by ivantodorovich** [comment]
@sergiocorato I only approved the github workflow run, not the actual PR -- I didn't do a review

**Comment by sergiocorato** [comment]
> @sergiocorato I only approved the github workflow run, not the actual PR -- I didn't do a review

Ah, ok, thanks

File: mrp_workorder_subcontracting/models/mrp_workorder.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+        if self.subcontracting_flow == "finished":
+            return self._get_subcontract_delivery_done_qty()
+
+        delivery_moves = self.delivery_move_ids.filtered(
+            lambda move: move.state != "cancel"
+        )
+        ratios = []
+        for product in delivery_moves.mapped("product_id"):
+            product_moves = delivery_moves.filtered(
+                lambda move, product=product: move.product_id == product
```
Comment (sergiocorato) [inline mrp_workorder_subcontracting/models/mrp_workorder.py]: To avoid confusion:
```suggestion
                lambda move, p=product: move.product_id == p
```

File: mrp_workorder_subcontracting/models/mrp_workorder.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            )
+            planned_qty = sum(product_moves.mapped("product_uom_qty"))
```
Comment (sergiocorato) [inline mrp_workorder_subcontracting/models/mrp_workorder.py]: product_moves could be in different uom, even for the same product.
This sum if not aware of it and could be wrong.

File: mrp_workorder_subcontracting/models/mrp_workorder.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+            done_qty = sum(
+                move.product_uom._compute_quantity(
```
Comment (sergiocorato) [inline mrp_workorder_subcontracting/models/mrp_workorder.py:434]: Here the conversion is from and to the same uom, I don't understand the reason.

**Comment by chrisandrewmann** [comment]
Thanks for the PR @aleboldro02 
I built something previously to fill a gap and was thinking about submitting it to OCA, so I wonder if yours also solves my problem?
I have clients who need to subcontract just the labour (service) during manufacturing work orders, such as engineer-to-order products that have to be sent for laser-engraving or CNC work, then returned to complete manufacturing onsite during a single MO.
This currently involves the MO creating a PO for each service and avoids any stock involvement, it's purely based on the work order triggering the PO for the service and has the service product linked to the work centre (subcontractor).

File: mrp_workorder_subcontracting/views/purchase_order_views.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="purchase_order_view_form_subcontracting" model="ir.ui.view">
+        <field name="name">purchase.order.form.subcontracting</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//div[@name='button_box']" position="inside">
```
Comment (sergiocorato) [inline mrp_workorder_subcontracting/views/purchase_order_views.xml:8]: A smart button to open the picking should be useful:

Image

File: mrp_workorder_subcontracting/views/mrp_workorder_views.xml [hunk 5]
```diff
@@ -0,0 +1,193 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="mrp_workorder_view_form_subcontracting" model="ir.ui.view">
+        <field name="name">mrp_workorder-form_inherit</field>
+        <field name="model">mrp.workorder</field>
+        <field name="inherit_id" ref="mrp.mrp_production_workorder_form_view_inherit" />
+        <field name="arch" type="xml">
+            <xpath expr="//page[@name='components']" position="after">
```
Comment (sergiocorato) [inline mrp_workorder_subcontracting/views/mrp_workorder_views.xml:38]: A button to create subcontract orders should be useful here:

Image

Comment (sergiocorato) [reply mrp_workorder_subcontracting/views/mrp_workorder_views.xml:38]: Also a smart button for PO created:

Image

File: mrp_workorder_subcontracting/models/mrp_workorder.py (hunk 2 above)
Comment (aleboldro02) [reply mrp_workorder_subcontracting/models/mrp_workorder.py]: Fixed. The computation now groups delivery moves by product and converts both planned and done quantities to the most precise UoM used by that product before computing the ratio.

File: mrp_workorder_subcontracting/models/mrp_workorder.py (hunk 3 above)
Comment (aleboldro02) [reply mrp_workorder_subcontracting/models/mrp_workorder.py:434]: Fixed together with the UoM issue above.

File: mrp_workorder_subcontracting/views/purchase_order_views.xml (hunk 4 above)
Comment (aleboldro02) [reply mrp_workorder_subcontracting/views/purchase_order_views.xml:8]: Added. A smart button is now available to open the related pickings from purchase order.

File: mrp_workorder_subcontracting/views/mrp_workorder_views.xml (hunk 5 above)
Comment (aleboldro02) [reply mrp_workorder_subcontracting/views/mrp_workorder_views.xml:38]: Added 3 button on mrp.workoder form view:
- Button to create subcontract orders
- Button to view purchase orders
- Button to view stock movements

**Comment by aleboldro02** [comment]
> Thanks for the PR @aleboldro02 I built something previously to fill a gap and was thinking about submitting it to OCA, so I wonder if yours also solves my problem? I have clients who need to subcontract just the labour (service) during manufacturing work orders, such as engineer-to-order products that have to be sent for laser-engraving or CNC work, then returned to complete manufacturing onsite during a single MO. This currently involves the MO creating a PO for each service and avoids any stock involvement, it's purely based on the work order triggering the PO for the service and has the service product linked to the work centre (subcontractor).

Hi @chrisandrewmann thanks for the comment!
If you need to manage the shipment of components along with the return of the finished product (production order product), then this module is designed for that!
When you close the last return, the system automatically complete the phase.
Based on the bill of materials configuration, the module allows you to send parts using real locations or the finished product (production order product) using virtual locations.
This way, if you configure inventory by contact, you’ll know exactly how many products have been shipped or are to be received from a subcontractor.

**The module is not designed to not handle inventory movements**.
This could be a starting point for a possible future extension of the module.
I don’t understand what you mean by “to complete manufacturing onsite during a single MO”...

**Comment by chrisandrewmann** [comment]
> > Thanks for the PR @aleboldro02 I built something previously to fill a gap and was thinking about submitting it to OCA, so I wonder if yours also solves my problem? I have clients who need to subcontract just the labour (service) during manufacturing work orders, such as engineer-to-order products that have to be sent for laser-engraving or CNC work, then returned to complete manufacturing onsite during a single MO. This currently involves the MO creating a PO for each service and avoids any stock involvement, it's purely based on the work order triggering the PO for the service and has the service product linked to the work centre (subcontractor).
> 
> Hi @chrisandrewmann thanks for the comment! If you need to manage the shipment of components along with the return of the finished product (production order product), then this module is designed for that! When you close the last return, the system automatically complete the phase. Based on the bill of materials configuration, the module allows you to send parts using real locations or the finished product (production order product) using virtual locations. This way, if you configure inventory by contact, you’ll know exactly how many products have been shipped or are to be received from a subcontractor.
> 
> **The module is not designed to not handle inventory movements**. This could be a starting point for a possible future extension of the module. I don’t understand what you mean by “to complete manufacturing onsite during a single MO”...

Thanks for the message.
It's hard to get across but basically it's the ability to out-source the work order operations to subcontractors, without having an impact on physical stock moves.
Using an MO there are multiple work orders example:

1. Assembly (internal)
2. Laser cutting (external subcontractor) - This work order triggers a PO for subcontracted service product
3. Testing
4. Packing

**Comment by aleboldro02** [comment]
> > > Thanks for the PR @aleboldro02 I built something previously to fill a gap and was thinking about submitting it to OCA, so I wonder if yours also solves my problem? I have clients who need to subcontract just the labour (service) during manufacturing work orders, such as engineer-to-order products that have to be sent for laser-engraving or CNC work, then returned to complete manufacturing onsite during a single MO. This currently involves the MO creating a PO for each service and avoids any stock involvement, it's purely based on the work order triggering the PO for the service and has the service product linked to the work centre (subcontractor).
> > 
> > 
> > Hi @chrisandrewmann thanks for the comment! If you need to manage the shipment of components along with the return of the finished product (production order product), then this module is designed for that! When you close the last return, the system automatically complete the phase. Based on the bill of materials configuration, the module allows you to send parts using real locations or the finished product (production order product) using virtual locations. This way, if you configure inventory by contact, you’ll know exactly how many products have been shipped or are to be received from a subcontractor.
> > **The module is not designed to not handle inventory movements**. This could be a starting point for a possible future extension of the module. I don’t understand what you mean by “to complete manufacturing onsite during a single MO”...
> 
> Thanks for the message. It's hard to get across but basically it's the ability to out-source the work order operations to subcontractors, without having an impact on physical stock moves. Using an MO there are multiple work orders example:
> 
> 1. Assembly (internal)
> 2. Laser cutting (external subcontractor) - This work order triggers a PO for subcontracted service product
> 3. Testing
> 4. Packing

A valid alternative is to send the finished product so that you can manage inventory using virtual locations without any actual physical movements.

If your only requirement is to generate the PO from the phase and not manage anything else:
- No inventory movements
- No automatic phase completion
Then this version of the module is NOT right for you.

**Comment by aleboldro02** [comment]
@OCA/manufacturing-maintainers could you take a look at this PR?

**Comment by aleboldro02** [comment]
Hi @LoisRForgeFlow , could you have look at this pull request?
This is a huge improvement that will make life easier for companies that outsource work orders on a daily basis without being able to predict them based on the bill of materials.
